Checkpoint - can now create a Survey (barebone, no associations yet)

coffeestrike opened the pull request.

coffeestrike committed:
Models for creating survey

- Survey model (title, description)
- Question model (description)
- Answer model (description, key_value_pairs)
- ResultPage model (title, description, image)

coffeestrike committed:
Included front end & qol gems

- Slim
- Simple_form
  rails generate simple_form:install --bootstrap
- Bootstrap(CDN)

coffeestrike committed:
Check bootstrap

- check bootstrap
- association for resultspage belongs_to survey

coffeestrike committed:
Can now create survey model (title & desc)

- changed some views to slim format (experimenting)
- created some controller methods for Survey

coffeestrike committed:
minor changes to survey controller

- added update
- added strongparams for nested attrs
- add accepts_nested_attr for survey model

coffeestrike committed:
Can now create survey which has many questions

- included gem cocoon

- the surveys create controller uses build to build questions based on its own id.

- in question model, belongs_to changed to optional as the nested form attemps to create it all at once and rails5 makes the belongs_to a compulsory.

coffeestrike committed:
Can now create a survey with questions and answers

coffeestrike committed:
Remove conflicts in gemfile

coffeestrike committed:
removed conflicts in gemfile lock

coffeestrike committed:
added gem cocoon

coffeestrike committed:
Start using RSpec

- gem rspec-rails & factory_girl_rails
- rails g rspec:install
- wrote a model test for Survey

- removed some conflict syntax in schema

coffeestrike committed:
fixed survey rspec to work

coffeestrike committed:
included factory girl & edited survey spec test

coffeestrike committed:
More Rspec tests & model changes

Created rspec model tests for each model
- has a valid factory
- can be created in db
- has validations

Created 3 factories for each model

Included gem
- rails-controller-testing
- shoulda-matchers

Stricter validations for each model

Added user_id foreign key for Survey

pynkvalerian commented on app/assets/javascripts/surveys.coffee:
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
remove file since it's unused.

pynkvalerian commented on app/assets/stylesheets/surveys.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Surveys controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
remove file since it's unused.

pynkvalerian commented on app/helpers/surveys_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module SurveysHelper
+end
```
remove file since it's unused.

pynkvalerian commented on app/models/answer.rb:
```diff
@@ -0,0 +1,7 @@
+class Answer < ApplicationRecord
+    belongs_to :question, optional: true
```
Why optional true but you validates presence of question_id?

pynkvalerian commented on app/models/survey.rb:
```diff
@@ -0,0 +1,9 @@
+class Survey < ApplicationRecord
+    belongs_to :user, optional: true
```
why optional true?

pynkvalerian commented on app/views/surveys/index.html.erb:
```diff
@@ -0,0 +1,2 @@
+<h1>Surveys#index</h1>
+<p>Find me in app/views/surveys/index.html.erb</p>
```
remove file since it's unused.

pynkvalerian commented on test/controllers/surveys_controller_test.rb:
```diff
@@ -0,0 +1,24 @@
+require 'test_helper'
```
remove all "test" folder since you're using rspec

pynkvalerian commented on db/migrate/20170203074044_create_surveys.rb:
```diff
@@ -0,0 +1,10 @@
+  def change
+    create_table :surveys do |t|
+      t.string :title
+      t.string :description
```
It'll be good to validate columns on both the model level (which you've already done) AND at the database level. 
Validating at the *model* level: when user submits form, validation occurs without hitting the database. 
Validating at *database* level: ensures integrity of data. 
ie for Survey, you `validates_presence_of :title and :description`, do include `null: false` for title and description columns as well.

pynkvalerian reviewed (changes requested):
Minor changes required.

coffeestrike committed:
Made changes as requested

1) Removed unnecessary files
- test files (since using Rspec)
- unused helpers
- unused scss

2) Edits to Survey#new slim files
- applied bootstrap col-md-x to differentiate Survey - Question - Answer forms.

pynkvalerian commented on app/controllers/surveys_controller.rb:
```diff
@@ -9,21 +9,12 @@ def show
     @survey = Survey.new(survey_params)
-    @survey.questions.build
-
+    byebug
```
byebug

coffeestrike committed:
Removed byebug in controller & null:false on model validations
